crates/sui/src/client_ptb/builder.rs

```diff
@@ -0,0 +1,993 @@
+struct NoResolution;
+
+#[async_trait]
+impl<'a> Resolver<'a> for NoResolution {
+    async fn resolve_object_id(
+        &mut self,
+        _builder: &mut PTBBuilder<'a>,
+        loc: Span,
+        _x: ObjectID,
+    ) -> PTBResult<Tx::Argument> {
+        error!(loc, "Don't resolve arguments and that's fine");
+    }
+}
+
+// ===========================================================================
+// PTB Builder and PTB Creation
+// ===========================================================================
+
+/// The PTBBuilder struct is the main workhorse that transforms a sequence of `ParsedPTBCommand`s
+/// into an actual PTB that can be run. The main things to keep in mind are that this contains:
+/// - A way to handle identifiers -- note that we "lazily" resolve identifiers to arguments, so
+///   that the first usage of the identifier determines what it is resolved to. If an identifier is
+///   used in multiple positions at different resolutions (e.g., in one place as an object argument,
+///   and in another as a pure value), this will result in an error. This error can be avoided by
+///   creating another identifier for the second usage.
+/// - A way to resolve arguments -- this is done by calling `resolve` on a `PTBArg` and passing in
+///   appropriate context. The context is used to determine how to resolve the argument -- e.g., if
+///   an object ID should be resolved to a pure value or an object argument.
+/// - A way to bind the result of a command to an identifier.
+pub struct PTBBuilder<'a> {
+    /// A map from identifiers to addresses. This is used to support address resolution, and also
+    /// supports external address sources (e.g., keystore).
+    addresses: BTreeMap<String, AccountAddress>,
+    /// A map from identifiers to the file scopes in which they were declared. This is used
+    /// for reporting shadowing warnings.
+    identifiers: BTreeMap<String, Vec<Span>>,
+    /// The arguments that we need to resolve. This is a map from identifiers to the argument
+    /// values -- they haven't been resolved to a transaction argument yet.
+    arguments_to_resolve: BTreeMap<String, Spanned<PTBArg>>,
+    /// The arguments that we have resolved. This is a map from identifiers to the actual
+    /// transaction arguments.
+    resolved_arguments: BTreeMap<String, Tx::Argument>,
+    /// Read API for reading objects from chain. Needed for object resolution.
+    reader: &'a ReadApi,
+    /// The last command that we have added. This is used to support assignment commands.
+    last_command: Option<Tx::Argument>,
+    /// The actual PTB that we are building up.
+    ptb: ProgrammableTransactionBuilder,
+    /// The list of errors that we have built up while processing commands. We do not report errors
+    /// eagerly but instead wait until we have processed all commands to report any errors.
+    errors: Vec<PTBError>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum ResolvedAccess {
```

Comment:
This is part of the fun of handling address aliases from the keystore that can contain `.`s....

Basically, we do some pre-resolution in `resolve_variable_access`: if we see dotted accesses, and if it's obviously a non-result access (e.g., `foo.bar.baz`) then `ResolvedAccess::DottedString` is returned. Otherwise if it's ambiguous (e.g., `foo.0`) `ResolvedAccess::ResultAccess` is returned, we then look to see if `foo` resolves to a `Result`, and if so turn that into a `NestedResult`, otherwise, we then look for an address alias called `foo.0`.

Reply:
> This is part of the fun of handling address aliases from the keystore that can contain `.`s....

I see, that's all my fault. Sorry!